### PR TITLE
fix(date-picker, input-date-picker): Fix setting date for the correct range input

### PR DESCRIPTION
**Pull request by driskull**
**Related Issue:** #2547

## Summary

fix(date-picker, input-date-picker): Fix setting date for the correct range input. #2547

- Uses activeRange to determine which date is being set.
- getDaysDiff() was returning odd results.
- Fixes mutable prop warnings.
- Keeps component active when setting range between start and end.

### CODE REVIEW

**Comment by driskull** [comment]
@paulcpederson should the range allow the same date as shown below? I think maybe this could be another setting on the range?

Screen Shot 2021-08-06 at 12 27 22 PM

**Comment by driskull** [comment]
This may also break `proximitySelectionDisabled`. Do we need to support that?

**Comment by julio8a** [comment]
Yeah, range should allow you to select the same day.

**Comment by driskull** [comment]
Ok. I thought so. currently it doesn't let you do that but I think its kind of messed up.

**Comment by driskull** [comment]
@paulcpederson what are your thoughts on removing `proximitySelectionDisabled`?

**Comment by julio8a** [comment]
gotcha, yea, we should fix that to allow same date selection.

**Comment by driskull** [comment]
I'll have to meet with @paulcpederson about this. The current UX in the `calcite-date-picker` is a bit odd. It's hard to select a range between two dates next to each other and impossible to select the same date as a range.

**Comment by paulcpederson** [comment]
@driskull I think @rpanichakit9541 may know more about the original requirements for range selection as he implemented that peice.

**Comment by rpanichakit9541** [comment]
@driskull - I believe the "proximitySelectionDisabled" flag spawned from this comment here https://github.com/Esri/calcite-components/pull/1136#issuecomment-712330139 .  But if we prefer to remove it and just one one range selection mode, that sounds good to me.  If you want to meet or if I can assist with anything let me know.  Also @bstifle might be able to speak to things as well since he contributed a lot to the design and PR feedback.

Here's the original issue https://github.com/Esri/calcite-components/issues/541

**Comment by bstifle** [comment]
@rpanichakit9541 what precisely do you need feedback on visually? (being able to select the same day in a range seems acceptable)

**Comment by rpanichakit9541** [comment]
@bstifle - I agree, I think the not being to select the same day is a bug.  I just added you to the thread in case there were other questions regarding original requirements.

**Comment by driskull** [comment]
Besides not being able to set the same day, its really tough just to get the range correct. Especially since when you hover a range and click, it doesn't set to what it is telling you. For example:

test

I would expect the range to be what the blue hover is set to but it does the opposite.

I think outside of this issue we should fix:

- Setting same range date
- Get hover and setting date range in sync (as seen in gif)

For the `calcite-input-date-picker`, this PR now limits the range to only before or after dates.

input

File: src/components/calcite-date-picker/calcite-date-picker.tsx [hunk 1]
```diff
@@ -40,10 +40,10 @@ export class CalciteDatePicker {
   //
   //--------------------------------------------------------------------------
   /** Active range */
-  @Prop() activeRange?: "start" | "end" = "start";
+  @Prop() activeRange?: "start" | "end";
```
Comment (driskull) [inline src/components/calcite-date-picker/calcite-date-picker.tsx:43]: For the default date-picker without an input, we don't want it to be in "start" or "end". We could have a new mode for both or just leave it empty.

**Comment by bstifle** [comment]
yeah its trying to find the day nearest to one of the selections.

but definitely if you focus on the second input, it should ONLY be choosing the end date

then with that, if focusing on the first input, only selecting the start date

**Comment by driskull** [comment]
@paulcpederson @rpanichakit9541 Can you review the PR?

File: src/components/calcite-date-picker/calcite-date-picker.tsx [hunk 2]
```diff
@@ -441,35 +466,30 @@ export class CalciteDatePicker {
...
       }
-      this.start = dateToISO(date);
-      this.setStartAsDate(date, true);
-      this.activeStartDate = date;
+      this.setStartDate(date);
     } else if (!this.endAsDate) {
-      this.end = dateToISO(date);
-      this.setEndAsDate(date, true);
-      this.activeEndDate = date;
+      this.setEndDate(date);
```
Comment (rpanichakit9541) [inline src/components/calcite-date-picker/calcite-date-picker.tsx:473]: Nice refactor! 👍

**Review by paulcpederson — APPROVED** [review]

**Comment by paulcpederson** [comment]
LGTM

**Review by rpanichakit9541 — APPROVED** [review]
LGTM 👍

**Pull request merged**